// bandcamp-fetch/dist/cjs/lib/autocomplete/AutocompleteResultsParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Parse_js_1 = require("../utils/Parse.js");
class AutocompleteResultsParser {
    static parseTags(json) {
        if (Array.isArray(json)) {
            return json.map((tagData) => ({
                type: 'tag',
                count: tagData.count,
                value: tagData.norm_name,
                name: tagData.display_name
            }));
        }
        throw new Parse_js_1.ParseError('Failed to parse autocomplete tags: JSON invalid or missing \'matching_tags\'.', json);
    }
    static parseLocations(json) {
        if (Array.isArray(json?.results)) {
            const arr = json.results;
            return arr.map((locationData) => ({
                type: 'location',
                value: parseInt(locationData.id, 10),
                name: locationData.name,
                fullName: locationData.fullname
            }));
        }
        throw new Parse_js_1.ParseError('Failed to parse autocomplete locations: JSON invalid or missing \'results\'.', json);
    }
}
exports.default = AutocompleteResultsParser;
