// bandcamp-fetch/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BandcampFetch = exports.Limiter = exports.Fetcher = exports.Cache = exports.StreamAPI = exports.AutocompleteAPI = exports.SearchAPI = exports.TrackAPI = exports.TagAPI = exports.ShowAPI = exports.ImageAPI = exports.FanAPI = exports.DiscoveryAPI = exports.BandAPI = exports.ArticleAPI = exports.AlbumAPI = exports.BaseAPIWithImageSupport = exports.BaseAPI = void 0;
const BandcampFetch_js_1 = __importDefault(require("./lib/BandcampFetch.js"));
var BaseAPI_js_1 = require("./lib/common/BaseAPI.js");
Object.defineProperty(exports, "BaseAPI", { enumerable: true, get: function () { return __importDefault(BaseAPI_js_1).default; } });
var BaseAPIWithImageSupport_js_1 = require("./lib/common/BaseAPIWithImageSupport.js");
Object.defineProperty(exports, "BaseAPIWithImageSupport", { enumerable: true, get: function () { return __importDefault(BaseAPIWithImageSupport_js_1).default; } });
var AlbumAPI_js_1 = require("./lib/album/AlbumAPI.js");
Object.defineProperty(exports, "AlbumAPI", { enumerable: true, get: function () { return __importDefault(AlbumAPI_js_1).default; } });
var ArticleAPI_js_1 = require("./lib/article/ArticleAPI.js");
Object.defineProperty(exports, "ArticleAPI", { enumerable: true, get: function () { return __importDefault(ArticleAPI_js_1).default; } });
var BandAPI_js_1 = require("./lib/band/BandAPI.js");
Object.defineProperty(exports, "BandAPI", { enumerable: true, get: function () { return __importDefault(BandAPI_js_1).default; } });
var DiscoveryAPI_js_1 = require("./lib/discovery/DiscoveryAPI.js");
Object.defineProperty(exports, "DiscoveryAPI", { enumerable: true, get: function () { return __importDefault(DiscoveryAPI_js_1).default; } });
var FanAPI_js_1 = require("./lib/fan/FanAPI.js");
Object.defineProperty(exports, "FanAPI", { enumerable: true, get: function () { return __importDefault(FanAPI_js_1).default; } });
var ImageAPI_js_1 = require("./lib/image/ImageAPI.js");
Object.defineProperty(exports, "ImageAPI", { enumerable: true, get: function () { return __importDefault(ImageAPI_js_1).default; } });
var ShowAPI_js_1 = require("./lib/show/ShowAPI.js");
Object.defineProperty(exports, "ShowAPI", { enumerable: true, get: function () { return __importDefault(ShowAPI_js_1).default; } });
var TagAPI_js_1 = require("./lib/tag/TagAPI.js");
Object.defineProperty(exports, "TagAPI", { enumerable: true, get: function () { return __importDefault(TagAPI_js_1).default; } });
var TrackAPI_js_1 = require("./lib/track/TrackAPI.js");
Object.defineProperty(exports, "TrackAPI", { enumerable: true, get: function () { return __importDefault(TrackAPI_js_1).default; } });
var SearchAPI_js_1 = require("./lib/search/SearchAPI.js");
Object.defineProperty(exports, "SearchAPI", { enumerable: true, get: function () { return __importDefault(SearchAPI_js_1).default; } });
var AutocompleteAPI_js_1 = require("./lib/autocomplete/AutocompleteAPI.js");
Object.defineProperty(exports, "AutocompleteAPI", { enumerable: true, get: function () { return __importDefault(AutocompleteAPI_js_1).default; } });
var StreamAPI_js_1 = require("./lib/stream/StreamAPI.js");
Object.defineProperty(exports, "StreamAPI", { enumerable: true, get: function () { return __importDefault(StreamAPI_js_1).default; } });
__exportStar(require("./lib/common/BaseAPI.js"), exports);
__exportStar(require("./lib/common/BaseAPIWithImageSupport.js"), exports);
__exportStar(require("./lib/album/AlbumAPI.js"), exports);
__exportStar(require("./lib/discovery/DiscoveryAPI.js"), exports);
__exportStar(require("./lib/image/ImageAPI.js"), exports);
__exportStar(require("./lib/article/ArticleAPI.js"), exports);
__exportStar(require("./lib/band/BandAPI.js"), exports);
__exportStar(require("./lib/fan/FanAPI.js"), exports);
__exportStar(require("./lib/show/ShowAPI.js"), exports);
__exportStar(require("./lib/tag/TagAPI.js"), exports);
__exportStar(require("./lib/track/TrackAPI.js"), exports);
__exportStar(require("./lib/album/AlbumAPI.js"), exports);
__exportStar(require("./lib/search/SearchAPI.js"), exports);
__exportStar(require("./lib/autocomplete/AutocompleteAPI.js"), exports);
__exportStar(require("./lib/stream/StreamAPI.js"), exports);
__exportStar(require("./lib/types/Album.js"), exports);
__exportStar(require("./lib/types/Shirt.js"), exports);
__exportStar(require("./lib/types/Article.js"), exports);
__exportStar(require("./lib/types/Artist.js"), exports);
__exportStar(require("./lib/types/Discovery.js"), exports);
__exportStar(require("./lib/types/Fan.js"), exports);
__exportStar(require("./lib/types/Image.js"), exports);
__exportStar(require("./lib/types/Label.js"), exports);
__exportStar(require("./lib/types/MediaKind.js"), exports);
__exportStar(require("./lib/types/Show.js"), exports);
__exportStar(require("./lib/types/Tag.js"), exports);
__exportStar(require("./lib/types/Track.js"), exports);
__exportStar(require("./lib/types/UserKind.js"), exports);
__exportStar(require("./lib/types/Search.js"), exports);
__exportStar(require("./lib/types/Autocomplete.js"), exports);
var Cache_js_1 = require("./lib/utils/Cache.js");
Object.defineProperty(exports, "Cache", { enumerable: true, get: function () { return __importDefault(Cache_js_1).default; } });
__exportStar(require("./lib/utils/Cache.js"), exports);
var Fetcher_js_1 = require("./lib/utils/Fetcher.js");
Object.defineProperty(exports, "Fetcher", { enumerable: true, get: function () { return __importDefault(Fetcher_js_1).default; } });
__exportStar(require("./lib/utils/Fetcher.js"), exports);
var Limiter_js_1 = require("./lib/utils/Limiter.js");
Object.defineProperty(exports, "Limiter", { enumerable: true, get: function () { return __importDefault(Limiter_js_1).default; } });
var BandcampFetch_js_2 = require("./lib/BandcampFetch.js");
Object.defineProperty(exports, "BandcampFetch", { enumerable: true, get: function () { return __importDefault(BandcampFetch_js_2).default; } });
__exportStar(require("./lib/BandcampFetch.js"), exports);
exports.default = new BandcampFetch_js_1.default();
