// bandcamp-fetch/dist/cjs/lib/track/TrackAPI.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _LimiterTrackAPI_limiter;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimiterTrackAPI = void 0;
const BaseAPIWithImageSupport_js_1 = __importDefault(require("../common/BaseAPIWithImageSupport.js"));
const TrackInfoParser_js_1 = __importDefault(require("./TrackInfoParser.js"));
class TrackAPI extends BaseAPIWithImageSupport_js_1.default {
    async getInfo(params) {
        const imageConstants = await this.imageAPI.getConstants();
        const opts = {
            trackUrl: params.trackUrl,
            imageBaseUrl: imageConstants.baseUrl,
            albumImageFormat: await this.imageAPI.getFormat(params.albumImageFormat, 9),
            artistImageFormat: await this.imageAPI.getFormat(params.artistImageFormat, 21),
            includeRawData: params.includeRawData !== undefined ? params.includeRawData : false
        };
        const html = await this.fetch(params.trackUrl);
        return TrackInfoParser_js_1.default.parseInfo(html, opts);
    }
}
exports.default = TrackAPI;
class LimiterTrackAPI extends TrackAPI {
    constructor(params) {
        super(params);
        _LimiterTrackAPI_limiter.set(this, void 0);
        __classPrivateFieldSet(this, _LimiterTrackAPI_limiter, params.limiter, "f");
    }
    async getInfo(params) {
        return __classPrivateFieldGet(this, _LimiterTrackAPI_limiter, "f").schedule(() => super.getInfo(params));
    }
}
exports.LimiterTrackAPI = LimiterTrackAPI;
_LimiterTrackAPI_limiter = new WeakMap();
