// bandcamp-fetch/dist/cjs/lib/fan/FanCollectionParser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const FanItemsBaseParser_js_1 = __importDefault(require("./FanItemsBaseParser.js"));
const FanWishlistParser_js_1 = __importDefault(require("./FanWishlistParser.js"));
class FanCollectionParser extends FanItemsBaseParser_js_1.default {
    static parseCollectionFromPage(html, opts) {
        return this.parsePageItems(html, {
            ...opts,
            dataKey: 'collection',
            parseItemFn: this.parseCollectionItem
        });
    }
    static parseCollectionItem(data, opts, tracklists) {
        return FanWishlistParser_js_1.default.parseWishlistItem(data, opts, tracklists);
    }
    static parseCollectionFromContinuation(json, continuation, opts) {
        return this.parseContinuationItems(json, continuation, {
            ...opts,
            dataKey: 'items',
            parseItemFn: this.parseCollectionItem
        });
    }
}
exports.default = FanCollectionParser;
