// bandcamp-fetch/dist/cjs/lib/tag/TagAPI.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _LimiterTagAPI_limiter;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimiterTagAPI = void 0;
const Constants_js_1 = require("../utils/Constants.js");
const BaseAPIWithImageSupport_js_1 = __importDefault(require("../common/BaseAPIWithImageSupport.js"));
const TagListParser_js_1 = __importDefault(require("./TagListParser.js"));
const Fetcher_js_1 = require("../utils/Fetcher.js");
const RelatedTagsParser_js_1 = __importDefault(require("./RelatedTagsParser.js"));
class TagAPI extends BaseAPIWithImageSupport_js_1.default {
    async list() {
        const html = await this.fetch(Constants_js_1.URLS.SITE_URL);
        return TagListParser_js_1.default.parseTags(html);
    }
    async getRelated(params) {
        const payload = {
            combo: true,
            tag_names: params.tags,
            size: params.size || 20
        };
        const json = await this.fetch(Constants_js_1.URLS.RELATED_TAGS, true, Fetcher_js_1.FetchMethod.POST, payload);
        return RelatedTagsParser_js_1.default.parseRelatedTags(json);
    }
}
exports.default = TagAPI;
class LimiterTagAPI extends TagAPI {
    constructor(params) {
        super(params);
        _LimiterTagAPI_limiter.set(this, void 0);
        __classPrivateFieldSet(this, _LimiterTagAPI_limiter, params.limiter, "f");
    }
    async list() {
        return __classPrivateFieldGet(this, _LimiterTagAPI_limiter, "f").schedule(() => super.list());
    }
    async getRelated(params) {
        return __classPrivateFieldGet(this, _LimiterTagAPI_limiter, "f").schedule(() => super.getRelated(params));
    }
}
exports.LimiterTagAPI = LimiterTagAPI;
_LimiterTagAPI_limiter = new WeakMap();
