// bandcamp-fetch/dist/cjs/lib/search/SearchAPI.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _LimiterSearchAPI_limiter;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimiterSearchAPI = exports.SearchItemType = void 0;
const url_1 = require("url");
const Constants_js_1 = require("../utils/Constants.js");
const SearchResultsParser_js_1 = __importDefault(require("./SearchResultsParser.js"));
const BaseAPIWithImageSupport_js_1 = __importDefault(require("../common/BaseAPIWithImageSupport.js"));
var SearchItemType;
(function (SearchItemType) {
    SearchItemType["All"] = "All";
    SearchItemType["ArtistsAndLabels"] = "ArtistsAndLabels";
    SearchItemType["Albums"] = "Albums";
    SearchItemType["Tracks"] = "Tracks";
    SearchItemType["Fans"] = "Fans";
})(SearchItemType || (exports.SearchItemType = SearchItemType = {}));
class SearchAPI extends BaseAPIWithImageSupport_js_1.default {
    async all(params) {
        return this.search({ ...params, itemType: SearchItemType.All });
    }
    async artistsAndLabels(params) {
        return this.search({ ...params, itemType: SearchItemType.ArtistsAndLabels });
    }
    async albums(params) {
        return this.search({ ...params, itemType: SearchItemType.Albums });
    }
    async tracks(params) {
        return this.search({ ...params, itemType: SearchItemType.Tracks });
    }
    async fans(params) {
        return this.search({ ...params, itemType: SearchItemType.Fans });
    }
    async search(params) {
        const opts = {
            itemType: params.itemType || SearchItemType.All,
            albumImageFormat: await this.imageAPI.getFormat(params.albumImageFormat, 9),
            artistImageFormat: await this.imageAPI.getFormat(params.artistImageFormat, 21)
        };
        const html = await this.fetch(SearchAPI.getSearchUrl(params));
        return SearchResultsParser_js_1.default.parseResults(html, opts);
    }
    /**
     * @internal
     */
    static getSearchUrl(params) {
        const urlObj = new url_1.URL(Constants_js_1.URLS.SEARCH);
        urlObj.searchParams.set('q', params.query);
        urlObj.searchParams.set('page', (params.page || 1).toString());
        switch (params.itemType) {
            case SearchItemType.ArtistsAndLabels:
                urlObj.searchParams.set('item_type', 'b');
                break;
            case SearchItemType.Albums:
                urlObj.searchParams.set('item_type', 'a');
                break;
            case SearchItemType.Tracks:
                urlObj.searchParams.set('item_type', 't');
                break;
            case SearchItemType.Fans:
                urlObj.searchParams.set('item_type', 'f');
                break;
            default:
        }
        return urlObj.toString();
    }
}
exports.default = SearchAPI;
class LimiterSearchAPI extends SearchAPI {
    constructor(params) {
        super(params);
        _LimiterSearchAPI_limiter.set(this, void 0);
        __classPrivateFieldSet(this, _LimiterSearchAPI_limiter, params.limiter, "f");
    }
    async all(params) {
        return __classPrivateFieldGet(this, _LimiterSearchAPI_limiter, "f").schedule(() => super.all(params));
    }
    async artistsAndLabels(params) {
        return __classPrivateFieldGet(this, _LimiterSearchAPI_limiter, "f").schedule(() => super.artistsAndLabels(params));
    }
    async albums(params) {
        return __classPrivateFieldGet(this, _LimiterSearchAPI_limiter, "f").schedule(() => super.albums(params));
    }
    async tracks(params) {
        return __classPrivateFieldGet(this, _LimiterSearchAPI_limiter, "f").schedule(() => super.tracks(params));
    }
    async fans(params) {
        return __classPrivateFieldGet(this, _LimiterSearchAPI_limiter, "f").schedule(() => super.fans(params));
    }
}
exports.LimiterSearchAPI = LimiterSearchAPI;
_LimiterSearchAPI_limiter = new WeakMap();
